src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultModelValidatorProvider.cs

```diff
@@ -21,7 +21,16 @@ public void GetValidators(ModelValidatorProviderContext context)
                 var validator = context.ValidatorMetadata[i] as IModelValidator;
                 if (validator != null)
                 {
-                    context.Validators.Add(validator);
+                    // Inserts validators based on whether or not they are 'required'. We want to run
+                    // 'required' validators first so that we get the best possible error message.
+                    if (validator.IsRequired)
+                    {
+                        context.Validators.Insert(0, validator);
```

Comment:
I was aware that it will not maintain the order. But I didn't think that maintaining the order was important.